lib/drivers/bayerContourNext.js

```diff
@@ -271,110 +287,139 @@ module.exports = function (config) {
 
   var bcnCommandResponse = function (commandpacket, callback) {
+    hidDevice.send(commandpacket, function () {
+      getASTMMessage(5000, 3, function(err) {
+        if (err !== null || !hidDevice.hasAvailablePacket()) {
+          callback(err, null);
+        } else {
+          // We're using the hidDevice as a message buffer, not a USB packet buffer
+          // to save on buffer copying. It could be refactored differently easily enough.
+          var message = hidDevice.nextPacket();
 
-    hidDevice.send(commandpacket.packet, function () {
-        receivePacket(5000, commandpacket.parser, function(err, packet) {
-            var rawtext = '';
-            while(hidDevice.hasAvailablePacket()){
-                var pkt = hidDevice.nextPacket();
-                pkt.parsed_payload = commandpacket.parser(pkt);
-                rawtext += pkt.parsed_payload;
-            }
-
-            if (err !== 'TIMEOUT') {
-              callback(err, rawtext);
-            }else{
-              debug('receivePacket timed out');
-              callback(err, null);
-            }
-        });
+          try {
+            callback(null, decodeMessage(message.payload));
+          }
+          catch(err) {
+            callback(err, null);
+          }
+        }
+      });
     });
   };
 
-  var receivePacket = function (timeout, parser, callback) {
+  var getASTMMessage = function (timeout, retries, callback) {
     var abortTimer = setTimeout(function () {
       clearInterval(listenTimer);
       debug('TIMEOUT');
-      callback('TIMEOUT', null);
+      var e = new Error('Timeout error.');
+      e.name = 'TIMEOUT';
+      callback(e, null);
     }, timeout);
 
     var listenTimer = setInterval(function () {
       hidDevice.receive(function(raw) {
-          var r = new Uint8Array(raw);
-          var startIndex = 0;
-          var rec = struct.unpack(r, 6, 'b', ['TYPE']);
-          var hasETB = _.indexOf(r, ASCII_CONTROL.ETB, startIndex);
-          if (hasETB !== -1) {
-            clearTimeout(abortTimer);
-          }
-          debug(rec);
-          if (hasETB !== -1) {
-            // send a new ACK
-            var cmd = buildAckPacket();
-            hidDevice.send(cmd, function () {
-                // we need to re-set timeout for each packet, as the number of packets are of variable length
-                abortTimer = setTimeout(function () {
-                  clearInterval(listenTimer);
-                  debug('TIMEOUT');
-                  callback('TIMEOUT', null);
-                }, timeout);
-                debug('New ACK SENT');
-            });
-          }
-          else if(rec.TYPE == ASCII_CONTROL.TER){
+        var packet = new Uint8Array(raw);
+
+        var packetHead = struct.unpack(packet, 0, '3Z2b', ['HEADER', 'SIZE', 'BYTE1']);
+
+        if(packetHead['HEADER'] !== MAGIC_HEADER){
+          debug('Invalid packet from Contour device');
+          callback(new Error('Invalid USB packet received.'), null);
```

Comment:
I was just testing the error handling by unplugging the device during an upload. You need to return the callback, otherwise it will still try to process and upload the invalid data. Also, remember to clear the timeout timer and interval timers.